src/Planner/CollectTableExpressionData.cpp

```diff
@@ -33,6 +35,25 @@ class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitor<CollectSource
 
     void visitImpl(QueryTreeNodePtr & node)
     {
+        /// Special case for USING clause which contains references to ALIAS columns.
+        /// We can not modify such ColumnNode.
+        if (auto * join_node = node->as<JoinNode>())
+        {
+            if (!join_node->isUsingJoinExpression())
+                return;
+
+            auto & using_list = join_node->getJoinExpression()->as<ListNode&>();
+            for (auto & using_element : using_list)
+            {
+                auto & column_node = using_element->as<ColumnNode&>();
+                auto & columns_from_subtrees = column_node.getExpressionOrThrow()->as<ListNode&>().getNodes();
+
+                visitUsingColumn(columns_from_subtrees[0]);
+                visitUsingColumn(columns_from_subtrees[1]);
```

Comment:
Can you add a comment here?
Why there are always 2 nodes, what they represent?

Reply:
Columns from left and right table. I'll add a comment.